manticore/core/executor.py

```diff
@@ -427,23 +427,26 @@ def run(self):
 
 
             while not self.is_shutdown():
-                try:
-                    #select a suitable state to analyze
-                    if current_state is None:
-                        with self._lock:
-                            #notify siblings we are about to stop this run
-                            self._notify_stop_run()
-                            #Select a single state_id
-                            current_state_id = self.get()
-                            #load selected state from secondary storage
-                            if current_state_id is not None:
-                                self._publish('will_load_state', current_state_id)
-                                current_state = self._workspace.load_state(current_state_id)
-                                self.forward_events_from(current_state, True)
-                                self._publish('did_load_state', current_state, current_state_id)
-                                logger.info("load state %r", current_state_id)
-                            #notify siblings we have a state to play with
-                            self._notify_start_run()
+                try:  # handle fatal errors: exceptions in Manticore
+                    try:  # handle external (e.g. solver) errors, and executor control exceptions
+                        #select a suitable state to analyze
+                        if current_state is None:
```

Comment:
what exceptions were you experiencing in this code? Concretize/TerminateState/SolverException?

Reply:
Some failing exception from the event callback. Any exception would to break it.

Reply:
wouldn't those still be caught by the bottom except block for Exceptions, AssertionError? moving the try up here doesn't do anything unless the exception from this block is a Concretize/TerminateState/SolverException?

Reply:
The new refactor _hopefully_ maintain the  
`
#{self._notify_start_run()} = #{self._notify_stop_run()}
`
invariant, even if exceptions are raised from the callbacks.